File: Classical Problems/Producer-Consumer/producer-consumer.py
import threading
import time
import random
import sys

buffer = [] # Items buffer (just a list)
buffer_lock = threading.Lock()
item_produced = threading.Condition()

def Producer(id):
    global buffer
    global buffer_lock
    global item_produced
    # Producer pseudo (from textbook):
    # event = waitForEvent()
    # buffer.add(event)
    print(f"producer {id} starts")
    for _ in range(3):
        added = random.randint(0,1)
        with buffer_lock:
            print(f"producer {id} adds {added} to the buffer")
            buffer.append(added)
        with item_produced:
            print(f"producer {id} notifies consumers they can consume")
            item_produced.notify_all()
        
    

def Consumer(id):
    global buffer
    global buffer_lock
    global item_produced
    # Consumer pseudo (from textbook):
    # event = buffer.pop()
    # event.process() (do something with event)
    print(f"consumer {id} starts")
    
    with item_produced:
        print(f"consumer {id} starts waiting for an item to be produced")
        
        
        with buffer_lock:
            while len(buffer) == 0:
                buffer_lock.release()
                item_produced.wait()
                print(f"consumer {id} returns from waiting")
                buffer_lock.acquire()
            
            item = buffer.pop()
            print(f"consumer {id} reads {item} from buffer")
    


if __name__=="__main__":
    # Threads must have exclusive access to buffer
    # If a consumer arrives while the buffer is empty, it blocks until
    # a producer has filled the thread
    
    for i in range(3):
        ci = threading.Thread(target=Consumer, args=(i,))
        ci.start()

    pi = threading.Thread(target=Producer, args=(0,))
    pi.start()
    
    # Cleanup
    parent = threading.current_thread()
    for thread in threading.enumerate():
        if thread is not parent:
            thread.join()

File: Classical Problems/Producer-Consumer/producer-consumer.rs
use std::thread;
use std::time::Duration;
use std::sync::Condvar;
use std::sync::Mutex;
use std::sync::MutexGuard;
use std::sync::Arc;

const NUM_PRODUCERS: i32 = 1;
const NUM_CONSUMERS: i32 = 10;

fn main() {
    // Track number of threads
    let mut threads = vec![];

    // Need a shared buffer
    let buffer = Arc::new(Mutex::new(vec![0]));

    // Item production condition variable
    let item_produced = Arc::new(Condvar::new());

    for i in 0..NUM_CONSUMERS {
        let cloned_buffer = buffer.clone();
        let cloned_item_produced = item_produced.clone();
        // Consumers spawner
        threads.push(thread::spawn(move || {
            // Consumer code
            println!("consumer {} starts", i);
            
            // Lock buffer and wait until consumable
            let buff_guard = cloned_buffer.lock().unwrap();
            // How do I actually surround this with a loop because of all these unwraps?
            //while buff_guard.len() == 0 { // <-- does not work
            let item = cloned_item_produced.wait(buff_guard).unwrap().pop();
            println!("consumer {} is done waiting, read {:?}",i,item);
            //}


            println!("consumer {} exits", i);
        }));
    }

    for j in 0..NUM_PRODUCERS {
        let cloned_buffer = buffer.clone();
        let cloned_item_produced = item_produced.clone();
        // Producer spawner
        threads.push(thread::spawn(move || {
            // Producer code
            println!("producer {} starts", j);

            for _ in 0..NUM_CONSUMERS {
                println!("producer {} relocks and adds to buffer",j);

                cloned_buffer.lock().unwrap().push(0);
                // Unlock?
                
                println!("producer {} notifies all consumers",j);
                // Notify thread that it can consume
                cloned_item_produced.notify_all();
            }

            println!("producer {} exits", j);
        }));
    }

    for t in threads {
        let _ = t.join().unwrap();
    }
}